### PR TITLE
Fix #4297: handle type param reference

### LINKED ISSUE #4297: Spurious is instanceOf warning
As reported by @allanrenucci  in #4245, the following code will get a warning saying `Some[Int]` cannot be checked at runtime.

```Scala
class Test {
  def test[X <: Option[Int]](x: X) = x.isInstanceOf[Some[Int]]
}
```

Initial diagnosis shows that type inference somehow failed to infer `a <: Int`, given `Some[a] <: X` and `X <: Option[Int]`.

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #4297: handle type param reference

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala [hunk 1]
```diff
@@ -88,19 +88,35 @@ object Checkable {
       }
     }
 
+    def maximize(implicit ctx: Context) = new ApproximatingTypeMap {
+      def apply(tp: Type): Type = tp match {
+        case tp: TypeRef if tp.underlying.isInstanceOf[TypeBounds] =>
+          val lo = this(tp.info.loBound)
+          val hi = this(tp.info.hiBound)
+          if (variance == 0) range(lo, hi)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: The `range` method already does variance checks like this.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: Thanks 👍

**Comment by liufengyun** [comment]
test performance with #fast please

File: compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala [hunk 2]
```diff
@@ -116,7 +130,9 @@ object Checkable {
           case defn.ArrayOf(tpE)   => recur(tpE, tpT)
           case _                   => recur(defn.AnyType, tpT)
         }
-      case tpe: AppliedType     => isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState())
+      case tpe: AppliedType     =>
+        isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState()) ||
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: Here, you first try without stripping type parameters for performance reasons? Maybe add a comment

File: compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala [hunk 3]
```diff
@@ -88,19 +88,33 @@ object Checkable {
       }
     }
 
+    def stripTypeParam(implicit ctx: Context) = new ApproximatingTypeMap {
+      def apply(tp: Type): Type = tp match {
+        case tp: TypeRef if tp.underlying.isInstanceOf[TypeBounds] =>
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: Probably equivalent to `case tp: TypeRef if tp.symbol.isTypeParam`. Maybe clearer?

File: compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala [hunk 4]
```diff
@@ -88,19 +88,33 @@ object Checkable {
       }
     }
 
+    def stripTypeParam(implicit ctx: Context) = new ApproximatingTypeMap {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: Maybe add a comment: `/** Approximate type parameters depending on variance */`

File: compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala [hunk 5]
```diff
@@ -116,7 +130,9 @@ object Checkable {
           case defn.ArrayOf(tpE)   => recur(tpE, tpT)
           case _                   => recur(defn.AnyType, tpT)
         }
-      case tpe: AppliedType     => isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState())
+      case tpe: AppliedType     =>
+        isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState()) ||
+        isClassDetermined(stripTypeParam.apply(X), tpe)(ctx.fresh.setNewTyperState())
```
[the remark this answers is not in the document]
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: It seems scalac has problem with this syntax, however, I've changed the method so we don't need to call `apply` anymore.

File: compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala (hunk 3 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala]: `tp.symbol.isTypeParam` doesn't cover the case of type members, so I think it's better to keep the original version here.

**Pull request merged**